### PR TITLE
Stitched blobs support different sized component blobs

### PR DESCRIPTION
**Pull request by dharju**
Changed stitched blobs to work with blobs with different data content sizes.

Note: Still need to add serializeMetadataContentV3 into BlobIdTransformer, but regular stitched blob operation changes are ready to be looked at

### CODE REVIEW

File: ambry-messageformat/src/main/java/com.github.ambry.messageformat/MessageFormatRecord.java [hunk 1]
```diff
@@ -1469,6 +1471,87 @@ public static CompositeBlobInfo deserializeMetadataContentRecord(DataInputStream
...
+     * Deserialize a metadata content record from a stream.
+     * @param stream The stream to read the serialized record from.
+     * @param storeKeyFactory The factory to use for parsing keys in the serialized metadata content record.
+     * @return A {@link CompositeBlobInfo} object with the chunk size and list of keys from the record.
+     * @throws IOException
+     */
+    public static CompositeBlobInfo deserializeMetadataContentRecord(DataInputStream stream,
+        StoreKeyFactory storeKeyFactory) throws IOException {
+      List<Pair<StoreKey, Long>> keysAndContentSizes = new ArrayList<>();
+      stream.readLong(); //total size
```
Comment (jsjtzyy) [inline ambry-messageformat/src/main/java/com.github.ambry.messageformat/MessageFormatRecord.java]: I suggest passing the total size into `CompositeBlobInfo` constructor and compare it with the calculated total size (which is `last` in your code). If they don't match, throw an exception.

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java [hunk 2]
```diff
@@ -937,9 +943,9 @@ protected ByteBuffer filterChunkToRange(ByteBuffer buf) {
         buf.position(0);
         buf.limit(0);
       } else {
-        long startOffsetInThisChunk = chunkIndex == 0 ? resolvedByteRange.getStartOffset() % chunkSize : 0;
```
Comment (jsjtzyy) [inline ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java]: I wonder previously if `chunkIndex > 0`, why the `startOffsetInThisChunk` is 0 here?

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java [hunk 3]
```diff
@@ -937,9 +943,9 @@ protected ByteBuffer filterChunkToRange(ByteBuffer buf) {
...
+        long startOffsetInThisChunk = chunkIndex == 0 ? resolvedByteRange.getStartOffset() - offset : 0;
         long endOffsetInThisChunk =
-            chunkIndex == (numChunksTotal - 1) ? (resolvedByteRange.getEndOffset() % chunkSize) + 1 : chunkSize;
```
Comment (jsjtzyy) [inline ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java]: One more minor question: is `endOffsetInThisChunk` exclusive or inclusive?
Looks like it is exclusive, right?

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java [hunk 4]
```diff
@@ -984,14 +990,14 @@ boolean isComplete() {
...
-    private List<StoreKey> keys;
+    private List<CompositeBlobInfo.StoreKeyAndSizeAndOffset> keysAndSizesAndOffsets;
     private BlobProperties serverBlobProperties;
 
     /**
      * Construct a FirstGetChunk and initialize it with the {@link BlobId} of the overall operation.
      */
     FirstGetChunk() {
-      super(-1, blobId);
+      super(-1, new CompositeBlobInfo.StoreKeyAndSizeAndOffset(blobId, 0L, -1L));
```
Comment (jsjtzyy) [inline ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java]: It seems the `chunkSize` is initialized to `-1L` here and I don't see it gets populated again because you removed several codes regarding setting `chunkSize`.  Correct me if I am wrong.

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java [hunk 5]
```diff
@@ -1051,7 +1057,6 @@ protected void maybeProcessCallbacks() {
                   decryptCallbackResultInfo.result.getDecryptedUserMetadata().array());
             }
             totalSize = decryptCallbackResultInfo.result.getDecryptedBlobContent().remaining();
-            chunkSize = totalSize;
```
Comment (jsjtzyy) [inline ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java]: Similar to previous comment. Where the `chunkSize` is populated? I notice that `getKeysAndSizesAndOffsets` contains chunk size info, however, note that `filterChunkToRange` method directly uses variable `chunkSize`, which might not be initialized. Please double check code there.

Comment (jsjtzyy) [reply ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java]: Also, we can discuss offline about the chunk size of a simple blob.

File: ambry-router/src/main/java/com.github.ambry.router/PutOperation.java [hunk 6]
```diff
@@ -1604,12 +1606,14 @@ private void finalizeMetadataChunk() {
               passedInBlobProperties.isEncrypted(), passedInBlobProperties.getExternalAssetTag());
       if (isStitchOperation() || getNumDataChunks() > 1) {
         // values returned are in the right order as TreeMap returns them in key-order.
-        List<StoreKey> orderedChunkIdList = new ArrayList<>(indexToChunkIds.values());
-        buf = MetadataContentSerDe.serializeMetadataContent(intermediateChunkSize, getBlobSize(), orderedChunkIdList);
+        List<Pair<StoreKey, Long>> orderedChunkIdList = indexToChunkIds.values().stream().collect(
+            Collectors.toList());
+
+        buf = MetadataContentSerDe.serializeMetadataContentV3(getBlobSize(), orderedChunkIdList);
```
Comment (cgtz) [inline ambry-router/src/main/java/com.github.ambry.router/PutOperation.java]: We don't want to start writing in v3 before all frontends have been deployed with read support for metadata content v3. 

Perhaps we should add a router config for the metadata content version like we have for blob id version and then pass the version number into the `MetadataContentSerDe.serializeMetadataContent()` method

**Comment by dharju** [comment]
Will be adding more unit tests (specifically unit test for MessageFormatRecord) and incorporation into BlobIdTransformer

File: ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java [hunk 7]
```diff
@@ -23,9 +26,41 @@
...
+      return offset;
+    }
+
+    public long getSize() {
+      return size;
+    }
+  }
+
+
   private final int chunkSize;
```
[the remark this answers is not in the document]
Comment (dharju) [reply ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java]: Chunk size would still be used for deserializing V2 metadata content and keys is used so that getKeys() call doesn't have to parse the StoreKeyAndSizeAndOffset object every time it is called

File: ambry-messageformat/src/main/java/com.github.ambry.messageformat/MessageFormatRecord.java [hunk 8]
```diff
@@ -1469,6 +1471,87 @@ public static CompositeBlobInfo deserializeMetadataContentRecord(DataInputStream
...
+
+    /**
+     * Get the total size of the metadata content record.
+     * @param keySize The size of each key in bytes.
+     * @param numberOfKeys The total number of keys.
+     * @return The total size in bytes.
+     */
+    public static int getMetadataContentSize(int keySize, int numberOfKeys) {
+      return Version_Field_Size_In_Bytes + TOTAL_SIZE_FIELD_SIZE_IN_BYTES + NUM_OF_KEYS_FIELD_SIZE_IN_BYTES + (
+          numberOfKeys * keySize * SIZE_OF_BLOB_FIELD_IN_BYTES);
```
[the remark this answers is not in the document]
Comment (dharju) [reply ambry-messageformat/src/main/java/com.github.ambry.messageformat/MessageFormatRecord.java]: You're right, adding unit test to catch this

File: ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java [hunk 9]
```diff
@@ -37,6 +72,57 @@ public CompositeBlobInfo(int chunkSize, long totalSize, List<StoreKey> keys) {
...
+    for (StoreKey key : keys) {
+      offsets.add(last);
+      keysAndSizesAndOffsets.add(new StoreKeyAndSizeAndOffset(key, last, Math.min(chunkSize, totalSize - last)));
+      last += chunkSize;
+    }
+  }
+
+  /**
+   * Construct a {@link CompositeBlobInfo} object.
+   * @param keysAndContentSizes
```
[the remark this answers is not in the document]
Comment (dharju) [reply ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java]: added

File: ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java [hunk 10]
```diff
@@ -37,6 +72,57 @@ public CompositeBlobInfo(int chunkSize, long totalSize, List<StoreKey> keys) {
...
+    this.totalSize = last;
+  }
+
+  /**
+   * Returns the keys (along with their data content size and offset relative to the total
+   * composite blob) of the chunks related to the given byte range of the entire composite
+   * blob
+   * @param start inclusive starting byte index
+   * @param end inclusive ending byte index
+   * @return
```
[the remark this answers is not in the document]
Comment (dharju) [reply ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java]: added

File: ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java [hunk 11]
```diff
@@ -37,6 +72,57 @@ public CompositeBlobInfo(int chunkSize, long totalSize, List<StoreKey> keys) {
...
+  public List<StoreKeyAndSizeAndOffset> getStoreKeysInByteRange(long start, long end) {
+    if (end < start || start < 0L || end >= totalSize) {
+      throw new IllegalArgumentException("Bad input parameters, start="+start+" end="+end+" totalSize="+totalSize);
+    }
+    int idx = Collections.binarySearch(offsets, start);
+    //binarySearch returns -(insertion point) - 1 if not an exact match, which points to the index of
+    //the first element greater than the key, or list.size() if all elements in the list are
+    // less than the specified key.
+    if (idx < 0) {
+      idx = -idx-2; //points to element with offset closest to but less than 'start'
```
[the remark this answers is not in the document]
Comment (dharju) [reply ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java]: The first offset will always be '0', so idx will never be -1 (it would only be -1 if 'start' was less than the first offset)

File: ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java [hunk 12]
```diff
@@ -37,6 +72,57 @@ public CompositeBlobInfo(int chunkSize, long totalSize, List<StoreKey> keys) {
...
+   * @return
+   */
+  public List<StoreKeyAndSizeAndOffset> getStoreKeysInByteRange(long start, long end) {
+    if (end < start || start < 0L || end >= totalSize) {
+      throw new IllegalArgumentException("Bad input parameters, start="+start+" end="+end+" totalSize="+totalSize);
+    }
+    int idx = Collections.binarySearch(offsets, start);
+    //binarySearch returns -(insertion point) - 1 if not an exact match, which points to the index of
+    //the first element greater than the key, or list.size() if all elements in the list are
+    // less than the specified key.
```
[the remark this answers is not in the document]
Comment (dharju) [reply ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java]: sure, changed to 'start'

File: ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java [hunk 13]
```diff
@@ -37,6 +72,57 @@ public CompositeBlobInfo(int chunkSize, long totalSize, List<StoreKey> keys) {
...
+   * Returns the keys (along with their data content size and offset relative to the total
+   * composite blob) of the chunks related to the given byte range of the entire composite
+   * blob
+   * @param start inclusive starting byte index
+   * @param end inclusive ending byte index
+   * @return
+   */
+  public List<StoreKeyAndSizeAndOffset> getStoreKeysInByteRange(long start, long end) {
+    if (end < start || start < 0L || end >= totalSize) {
+      throw new IllegalArgumentException("Bad input parameters, start="+start+" end="+end+" totalSize="+totalSize);
```
[the remark this answers is not in the document]
Comment (dharju) [reply ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java]: done

File: ambry-messageformat/src/main/java/com.github.ambry.messageformat/MessageFormatRecord.java (hunk 1 above)
Comment (dharju) [reply ambry-messageformat/src/main/java/com.github.ambry.messageformat/MessageFormatRecord.java]: I'm not for adding the total size as a param to the `CompositeBlobInfo` ctor to just perform a check, but I will add the check here, and in `serializeMetadataContentRecord`

File: ambry-router/src/main/java/com.github.ambry.router/PutOperation.java [hunk 14]
```diff
@@ -1604,12 +1606,14 @@ private void finalizeMetadataChunk() {
               passedInBlobProperties.isEncrypted(), passedInBlobProperties.getExternalAssetTag());
       if (isStitchOperation() || getNumDataChunks() > 1) {
         // values returned are in the right order as TreeMap returns them in key-order.
-        List<StoreKey> orderedChunkIdList = new ArrayList<>(indexToChunkIds.values());
-        buf = MetadataContentSerDe.serializeMetadataContent(intermediateChunkSize, getBlobSize(), orderedChunkIdList);
+        List<Pair<StoreKey, Long>> orderedChunkIdList = indexToChunkIds.values().stream().collect(
+            Collectors.toList());
```
[the remark this answers is not in the document]
Comment (dharju) [reply ambry-router/src/main/java/com.github.ambry.router/PutOperation.java]: I don't think there's a specific reason, replaced

File: ambry-router/src/main/java/com.github.ambry.router/PutOperation.java [hunk 15]
```diff
@@ -1619,7 +1623,8 @@ private void finalizeMetadataChunk() {
      * @return a list of all of the successfully put chunk ids associated with this blob
      */
     List<StoreKey> getSuccessfullyPutChunkIds() {
-      return new ArrayList<>(indexToChunkIds.values());
+      return indexToChunkIds.values().stream().map(b -> b.getFirst()).collect(
+          Collectors.toList());
```
[the remark this answers is not in the document]
Comment (dharju) [reply ambry-router/src/main/java/com.github.ambry.router/PutOperation.java]: sure

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java (hunk 3 above)
Comment (dharju) [reply ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java]: exclusive since it's used in `Buffer.limit()`.  I can rename it `endOffsetInThisChunkExclusive`

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java (hunk 2 above)
Comment (dharju) [reply ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java]: because it's relative to the chunk itself. The chunkIndexes only refer to chunks that are within a byte range.  If the chunkIndex refers to a chunk beyond the first chunk, it will have to start at 0.

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java (hunk 4 above)
Comment (dharju) [reply ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java]: That's right, but that's ok since chunkSize is only used when calculating the end of data chunk for a byte range operation if that data chunk is not the last data chunk.  For simple data blobs, the FirstChunk will always be the last data chunk so it's unused, and for metadata blobs the size of the FirstChunk is irrelevant.

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java (hunk 5 above)
Comment (dharju) [reply ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java]: It's ok since chunkSize is only used when calculating the end of data chunk for a byte range operation if that data chunk is not the last data chunk. For simple data blobs, the FirstChunk will always be the last data chunk so it's unused, and for metadata blobs the size of the FirstChunk is irrelevant.

File: ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java [hunk 16]
```diff
@@ -23,9 +26,41 @@
  * total composite blob size, and a list of keys for the blob's data chunks.
  */
 public class CompositeBlobInfo {
+
+  /**
+   * POJO class for holding a store key, the size of the data content it refers to,
+   * and the offset of the data in the larger composite blob
+   */
+  public static class StoreKeyAndSizeAndOffset {
```
[the remark this answers is not in the document]
Comment (dharju) [reply ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java]: sure

File: ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java [hunk 17]
```diff
@@ -37,6 +72,57 @@ public CompositeBlobInfo(int chunkSize, long totalSize, List<StoreKey> keys) {
...
+      throw new IllegalArgumentException("Bad input parameters, start="+start+" end="+end+" totalSize="+totalSize);
+    }
+    int idx = Collections.binarySearch(offsets, start);
+    //binarySearch returns -(insertion point) - 1 if not an exact match, which points to the index of
+    //the first element greater than the key, or list.size() if all elements in the list are
+    // less than the specified key.
+    if (idx < 0) {
+      idx = -idx-2; //points to element with offset closest to but less than 'start'
+    }
+    List<StoreKeyAndSizeAndOffset> ans = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (dharju) [reply ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java]: We can, and another cool thing about using `subList()` is that I can do another binary search for the end index replacing the linearly scaling while loop and make this method a O(log(n)) operation (since `subList()` itself is a constant operation for `ArrayList`).  Great suggestion!

File: ambry-router/src/main/java/com.github.ambry.router/PutOperation.java (hunk 6 above)
Comment (dharju) [reply ambry-router/src/main/java/com.github.ambry.router/PutOperation.java]: added config

**Comment by dharju** [comment]
Fixing failing check, also changing PutManagerTest, GetManagerTest, and NonBlockingRouterTest to test against both old and new versions of metadata content

**Comment by dharju** [comment]
Looking at GetBlobOperationTest failure, appears to be triggered by random chunk size, indicates that a kind of corner case is missing from the test, will fix

**Comment by dharju** [comment]
Source of error was test with a very low probability of generating an incorrect test input.  Fixing

File: ambry-replication/src/test/java/com.github.ambry.replication/BlobIdTransformerTest.java [hunk 18]
```diff
@@ -155,15 +157,23 @@ public void testBasicOperation() throws Exception {
...
-    InputAndExpected inputAndExpected =
-        new InputAndExpected(BLOB_ID_VERSION_1_METADATA_CONVERTED, VALID_MESSAGE_FORMAT_INPUT_STREAM_IMPLS[0], false,
-            new String[]{BLOB_ID_VERSION_1_DATACHUNK_0_CONVERTED.getFirst(),
-                BLOB_ID_VERSION_1_DATACHUNK_1_CONVERTED.getFirst()},
-            new String[]{BLOB_ID_VERSION_1_DATACHUNK_0_CONVERTED.getSecond(),
-                BLOB_ID_VERSION_1_DATACHUNK_1_CONVERTED.getSecond()});
-    TransformationOutput output = transformer.transform(inputAndExpected.getInput());
-    assertNull("output exception should be null", output.getException());
-    verifyOutput(output.getMsg(), inputAndExpected.getExpected());
+    for (int i = 0; i < 2; i++) {
```
Comment (jsjtzyy) [inline ambry-replication/src/test/java/com.github.ambry.replication/BlobIdTransformerTest.java]: Move this piece of code to `BlobIdTransformerTest` ctor. Use parameters to test both V2 and V3 metadataContent in this class.

File: ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java [hunk 19]
```diff
@@ -62,4 +126,49 @@ public int getChunkSize() {
   public List<StoreKey> getKeys() {
```
Comment (cgtz) [inline ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java]: An alternative to materializing a list of `StoreKey`s for every `CompositeBlobInfo` is to return a "view" of `chunkMetadataList`:
```java
  public List<StoreKey> getKeys() {
    return new AbstractList<StoreKey>() {
      @Override
      public StoreKey get(int index) {
        return chunkMetadataList.get(index).getStoreKey();
      }

      @Override
      public int size() {
        return chunkMetadataList.size();
      }
    };
  }
```

It seems like this would suffice for the two use cases that need the `getKeys()` method, the `getChunkIdsOnly` option in the router and `BlobIdTransformer`, as both of them are just interested in iterating through all of the keys and not mutating the list.

File: ambry-messageformat/src/main/java/com.github.ambry.messageformat/MessageFormatRecord.java [hunk 20]
```diff
@@ -1469,6 +1471,97 @@ public static CompositeBlobInfo deserializeMetadataContentRecord(DataInputStream
       return new CompositeBlobInfo(chunkSize, totalSize, keys);
     }
   }
+
+  /**
+   *  - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+   * |         |              |            |            |          |            |            |          |
+   * | version |  total size  | # of keys  |  size of   |   key1   |  size of   |    key2    |  ......  |
+   * |(2 bytes)|  (8 bytes)   | (4 bytes)  | key1 blob  |          | key2 blob  |            |  ......  |
```
Comment (cgtz) [inline ambry-messageformat/src/main/java/com.github.ambry.messageformat/MessageFormatRecord.java]: Is the totalSize field meant for sanity checking that `sum(chunksizes) == totalsize` or are there other places where it is used?

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java [hunk 21]
```diff
@@ -552,11 +555,14 @@ void reset() {
     /**
      * Assign a chunk of the overall blob to this GetChunk.
      * @param index the index of the chunk of the overall blob that needs to be fetched through this GetChunk.
-     * @param id the id of the chunk of the overall blob that needs to be fetched through this GetChunk.
+     * @param keyAndSizeAndOffset the id, data content size, and offset (relative to the total size)
+     *                           of the chunk of the overall blob that needs to be fetched through this GetChunk.
      */
-    void initialize(int index, BlobId id) {
+    void initialize(int index, CompositeBlobInfo.ChunkMetadata keyAndSizeAndOffset) {
```
Comment (cgtz) [inline ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java]: change variable names in this class (and others) from `keyAndSizeAndOffset` to `chunkMetadata`

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java (hunk 2 above)
Comment (cgtz) [reply ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java]: Sorry about the confusion around `chunkIndex`. Since we now have the `offset` variable, you could consider changing this to `long startOffsetInThisChunk = Math.max(resolvedByteRange.getStartOffset() - offset, 0)` if you prefer.

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java (hunk 3 above)
Comment (cgtz) [reply ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java]: If you feel that it is cleaner, you could also express this as `long endOffsetInThisChunkExclusive = Math.min(resolvedByteRange.getEndOffset() - offset + 1, chunkSize)`

File: ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java [hunk 22]
```diff
@@ -37,6 +44,63 @@ public CompositeBlobInfo(int chunkSize, long totalSize, List<StoreKey> keys) {
...
+  /**
+   * Returns the keys (along with their data content size and offset relative to the total
+   * composite blob) of the chunks related to the given byte range of the entire composite
+   * blob
+   * @param start inclusive starting byte index
+   * @param end inclusive ending byte index
+   * @return the list of {@link ChunkMetadata} that lie upon the inclusive range
+   * between 'start' and 'end'
+   */
+  public List<ChunkMetadata> getStoreKeysInByteRange(long start, long end) {
```
[the remark this answers is not in the document]
Comment (dharju) [reply ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java]: agreed, actually writing that now.  `CompositeBlobInfo` didn't have a unit test before because it was a POJO class but with this method it should be tested.

[the remark this answers is not in the document]
Comment (dharju) [reply ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java]: Can you explain "0 byte chunks in the middle of the blob"?

File: ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java [hunk 23]
```diff
@@ -37,6 +44,63 @@ public CompositeBlobInfo(int chunkSize, long totalSize, List<StoreKey> keys) {
...
+    // less than 'start'.
+    if (startIdx < 0) {
+      startIdx = -startIdx - 2; //points to element with offset closest to but less than 'start'
+    }
+
+    int endIdx = Collections.binarySearch(offsets, end);
+    if (endIdx < 0) {
+      endIdx = -endIdx - 2; //points to element with offset closest to but less than 'end'
+    }
+    endIdx++;//List.subList expects an exclusive index
```
[the remark this answers is not in the document]
Comment (dharju) [reply ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java]: done

File: ambry-api/src/main/java/com.github.ambry/config/RouterConfig.java [hunk 24]
```diff
@@ -193,6 +193,13 @@
   @Default("5")
   public final short routerBlobidCurrentVersion;
 
+  /**
+   * The version to use for new metadata blobs.
+   */
+  @Config("router.metadata.content.version")
```
[the remark this answers is not in the document]
Comment (dharju) [reply ambry-api/src/main/java/com.github.ambry/config/RouterConfig.java]: I like `router.metadata.content.version` because it matches the terminology used in the `Metadata_Content_Format_*` classes

File: ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java [hunk 25]
```diff
@@ -62,4 +126,49 @@ public int getChunkSize() {
...
+   * key's data content size and offset relative to the total composite blob
+   * @return A list of {@link ChunkMetadata}
+   */
+  public List<ChunkMetadata> getChunkMetadataList() {
+    return chunkMetadataList;
+  }
+
+  /**
+   * Return the version of the metadata content record that this object was deserialized from
+   * @return
```
[the remark this answers is not in the document]
Comment (dharju) [reply ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java]: done

File: ambry-messageformat/src/test/java/com.github.ambry.messageformat/MessageFormatRecordTest.java [hunk 26]
```diff
@@ -507,6 +553,27 @@ private CompositeBlobInfo deserializeMetadataContentV2(ByteBuffer metadataConten
...
+        keysAndContentSizes.get(0).getFirst().sizeInBytes(), keysAndContentSizes.size());
+    ByteBuffer metadataContent = ByteBuffer.allocate(size);
+    MessageFormatRecord.Metadata_Content_Format_V3.serializeMetadataContentRecord(metadataContent, totalSize,
+        keysAndContentSizes);
+    metadataContent.flip();
+    return metadataContent;
+  }
+
+  private CompositeBlobInfo deserializeMetadataContentV3(ByteBuffer metadataContent, StoreKeyFactory storeKeyFactory)
+      throws MessageFormatException, IOException {
```
[the remark this answers is not in the document]
Comment (dharju) [reply ambry-messageformat/src/test/java/com.github.ambry.messageformat/MessageFormatRecordTest.java]: removed

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java (hunk 21 above)
Comment (dharju) [reply ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java]: done

File: ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (cgtz) [reply ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java]: Sure. So I was thinking that since we allow 0 byte simple blob uploads, they are valid candidates for stitching. So, someone could potentially create a blob with chunk sizes like: 25, 30, 0, 0, 70, 0, 5

If we call this method with a `start` offset of 30, would the range start from chunk 1 or 2?

[the remark this answers is not in the document]
Comment (dharju) [reply ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java]: As discussed in person, going to change the code so that stitched blobs will not allow blobs of 0 length to be stitched (will throw an exception if that happens)

File: ambry-messageformat/src/main/java/com.github.ambry.messageformat/MessageFormatRecord.java (hunk 20 above)
Comment (dharju) [reply ambry-messageformat/src/main/java/com.github.ambry.messageformat/MessageFormatRecord.java]: Yeah it's only used for checking

File: ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java (hunk 19 above)
Comment (dharju) [reply ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java]: Cool idea, I'll use it

**Comment by dharju** [comment]
made the max, min changes suggested above

File: ambry-replication/src/test/java/com.github.ambry.replication/BlobIdTransformerTest.java (hunk 18 above)
Comment (dharju) [reply ambry-replication/src/test/java/com.github.ambry.replication/BlobIdTransformerTest.java]: agreed, changed

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java [hunk 27]
```diff
@@ -943,8 +943,9 @@ protected ByteBuffer filterChunkToRange(ByteBuffer buf) {
         buf.position(0);
         buf.limit(0);
       } else {
-        long startOffsetInThisChunk = Math.max(resolvedByteRange.getStartOffset() - offset, 0);
-        long endOffsetInThisChunkExclusive = Math.min(resolvedByteRange.getEndOffset() - offset + 1, chunkSize);
+        long startOffsetInThisChunk = chunkIndex == 0 ? resolvedByteRange.getStartOffset() - offset : 0;
```
[the remark this answers is not in the document]
Comment (dharju) [reply ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java]: It broke a GetBlobOperationTest test so I changed it back

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java [hunk 28]
```diff
@@ -1241,11 +1243,13 @@ private void initializeDataChunks() {
         numChunksTotal = 0;
         dataChunks = null;
       } else {
-        chunkIdIterator = keys.listIterator();
-        numChunksTotal = keys.size();
-        dataChunks = new GetChunk[Math.min(keys.size(), NonBlockingRouter.MAX_IN_MEM_CHUNKS)];
+        chunkIdIterator = keysAndSizesAndOffsets.listIterator();
```
[the remark this answers is not in the document]
Comment (dharju) [reply ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java]: thought I caught them all... done

File: ambry-router/src/main/java/com.github.ambry.router/PutOperation.java [hunk 29]
```diff
@@ -1604,12 +1612,21 @@ private void finalizeMetadataChunk() {
               passedInBlobProperties.isEncrypted(), passedInBlobProperties.getExternalAssetTag());
       if (isStitchOperation() || getNumDataChunks() > 1) {
         // values returned are in the right order as TreeMap returns them in key-order.
-        List<StoreKey> orderedChunkIdList = new ArrayList<>(indexToChunkIds.values());
-        buf = MetadataContentSerDe.serializeMetadataContent(intermediateChunkSize, getBlobSize(), orderedChunkIdList);
+        List<Pair<StoreKey, Long>> orderedChunkIdList = new ArrayList<>(indexToChunkIdsAndChunkSizes.values());
```
Comment (cgtz) [inline ambry-router/src/main/java/com.github.ambry.router/PutOperation.java]: Move this to inside the else if (line 1620) so that we don't have to construct it for v2

Comment (dharju) [reply ambry-router/src/main/java/com.github.ambry.router/PutOperation.java]: done
